aip/0145.md

```diff
@@ -0,0 +1,65 @@
+---
+aip:
+  id: 145
+  state: reviewing
+  created: 2020-05-28
+permalink: /145
+redirect_from:
+  - /0145
+---
+
+# Ranges
+
+Services often need to represent ranges of discrete or continuous values. These
+have wide differences in meaning, and come in many types: integers, floats, and
+timestamps, just to name a few, and the expected meaning of a range can vary in
+subtle ways depending on the type of range being discussed.
+
+## Guidance
+
+A resource or message representing a range **should** ordinarily use two
+separate fields of the same type, with prefixes `start_` and `end_`:
+
+```proto
+// A representation of a chapter in a book.
+message Chapter {
+  string title = 1;
+
+  // The page where this chapter begins.
+  int32 start_page = 2;
+
+  // The page where the next chapter or section begins.
+  int32 end_page = 3;
+}
+```
+
+### Inclusive or exclusive ranges
+
+Fields representing ranges **should** use inclusive start values and exclusive
+end values (half-closed intervals) in most situations; in interval notation:
+`[start_xxx, end_xxx)`.
+
+Exclusive end values are preferable for the following reasons:
+
+- It conforms to user expectations, particularly for continuous values such as
+  timestamps, and avoids the need to express imprecise "limit values" (e.g.
+  `2012-04-20T23:59:59`).
+- It is consistent with most common programming languages, including C++, Java,
```

Comment:
I'd add the benefit of simple abutting ranges: `[0, x)`, [x, y)`, `[y, z)` where the only values are the split points. It's much easier to get that right can with inclusive end points.

Reply:
You meant exclusive end points. ;-)

Reply:
Done.